Mathlib/LinearAlgebra/PiTensorProduct.lean

```diff
@@ -463,17 +471,99 @@ then there is an induced map `⨂ᵢ sᵢ → ⨂ᵢ tᵢ` by `⨂ aᵢ ↦ ⨂
 
 This is `TensorProduct.map` for an arbitrary family of modules.
 -/
-def map (f : Π i, s i →ₗ[R] t i) : (⨂[R] i, s i) →ₗ[R] ⨂[R] i, t i :=
+def map : (⨂[R] i, s i) →ₗ[R] ⨂[R] i, t i :=
   lift <| (tprod R).compLinearMap f
 
-@[simp] lemma map_tprod (f : Π i, s i →ₗ[R] t i) (x : Π i, s i) :
+@[simp] lemma map_tprod (x : Π i, s i) :
     map f (tprod R x) = tprod R fun i ↦ f i (x i) :=
   lift.tprod _
 
+-- No lemmas about associativity, because we don't have associativity of `PiTensorProduct` yet.
+
+theorem map_range_eq_span_tprod : LinearMap.range (map f) =
+    Submodule.span R { t | ∃ (m : (i : ι) → s i), tprod R (fun i ↦ f i (m i)) = t } := by
+  simp only [← Submodule.map_top, ← span_tprod_eq_top, Submodule.map_span]
+  congr; ext t
+  simp only [Set.mem_image, Set.mem_setOf_eq, exists_exists_eq_and, map_tprod]
+
+/-- Given submodules `p i ⊆ s i`, this is the natural map: `⨂[R] i, p i → ⨂[R] i, s i`. -/
+@[simp]
+def mapIncl (p : (i : ι) → Submodule R (s i)) : (⨂[R] i, p i) →ₗ[R] ⨂[R] i, s i :=
+  map (fun (i : ι) ↦ (p i).subtype)
+
+theorem map_comp : map (fun (i : ι) ↦ (g i).comp (f i)) = (map g).comp (map f) := by
+  ext
+  simp only [LinearMap.compMultilinearMap_apply, map_tprod, LinearMap.coe_comp, Function.comp_apply]
+
+theorem lift_comp_map (h : MultilinearMap R t E) :
+    lift h ∘ₗ map f = lift (h.compLinearMap f) := by
+  ext
+  simp only [LinearMap.compMultilinearMap_apply, LinearMap.coe_comp, Function.comp_apply,
+    map_tprod, lift.tprod, MultilinearMap.compLinearMap_apply]
+
+attribute [local ext high] ext
+
+@[simp]
+theorem map_id : map (fun (i : ι) ↦ (LinearMap.id : s i →ₗ[R] s i))  = .id := by
+  ext
+  simp only [LinearMap.compMultilinearMap_apply, map_tprod, LinearMap.id_coe, id_eq]
+
+@[simp]
+theorem map_one : map (fun (i : ι) ↦ (1 : s i →ₗ[R] s i)) = 1 :=
+  map_id
+
+theorem map_mul (f₁ f₂ : Π i, s i →ₗ[R] s i)  :
+    map (fun (i : ι) ↦ (f₁ i) * (f₂ i)) = map f₁ * map f₂ :=
+  map_comp f₁ f₂
+
+@[simp]
+protected theorem map_pow (f : Π i, s i →ₗ[R] s i) (n : ℕ) : map f ^ n = map (f ^ n) := by
+  induction' n with n ih
+  · simp only [Nat.zero_eq, pow_zero]; rw [← map_one]; rfl
+  · simp only [pow_succ', ih]; rw [← map_mul]; rfl
```

Comment:
This can proved in one line if you show that `map` is a `RingHom`

Reply:
Okay, but then we should also prove the corresponding result for [`TensorProduct.map`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/TensorProduct.html#TensorProduct.map), to get a shorter proof of [`TensorProduct.map_pow`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/TensorProduct.html#TensorProduct.map_pow).

Reply:
Wait, now that the coffee kicked in, I have to ask: a `RingHom` from what to what ? Since `map` is multilinear in the `f i`, the only possibility I see is to prove that the lift of `mapMultilinear` is a `RingHom` from `(⨂[R] i, s i →ₗ[R] s i)` to `((⨂[R] i, s i) →ₗ[R] ⨂[R] i, s i)`, but I don't know if that's what you meant.

Reply:
I think I must have meant a MonoidHom

Reply:
Like so ? I now question whether we should have `map_pow` at all, since the prood is literally `MonoidHom.map_pow`.

Reply:
Indeed the linter is complaining that `map_pow` can now be proved by `simp`, and I agree with it for once.